File: app/src/main/java/com/chess/cryptobot/task/AvailablePairsTask.java
package com.chess.cryptobot.task;

import android.content.Context;

import com.chess.cryptobot.R;
import com.chess.cryptobot.content.ContextHolder;
import com.chess.cryptobot.content.pairs.PairsHolder;
import com.chess.cryptobot.exceptions.MarketException;
import com.chess.cryptobot.market.Market;
import com.chess.cryptobot.model.response.TickerResponse;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class AvailablePairsTask extends MarketTask<Integer, List<String>> {
    private List<String> availablePairNames;
    private Map<String, Double> bittrexVolumes;
    private Map<String, Double> livecoinVolumes;

    public AvailablePairsTask(PairsHolder pairsHolder) {
        super(pairsHolder);
    }

    @Override
    public void preMarketProcess(Integer param) {
        bittrexVolumes = new HashMap<>();
        livecoinVolumes = new HashMap<>();
    }

    @Override
    public List<String> marketProcess(Market market, Integer param) throws MarketException {
        List<String> pairNames;
        List<? extends TickerResponse> tickers = market.getTicker();
        pairNames = getPairNames(tickers);
        updateVolumesForMarket(market.getMarketName(), tickers);

        if (availablePairNames == null) {
            availablePairNames = new LinkedList<>(pairNames);
        } else {
            availablePairNames.retainAll(pairNames);
            livecoinVolumes.keySet().retainAll(availablePairNames);
            bittrexVolumes.keySet().retainAll(availablePairNames);
        }

        return availablePairNames;
    }

    private List<String> getPairNames(List<? extends TickerResponse> tickers) {
        List<String> pairNames = new ArrayList<>();
        tickers.forEach(ticker -> pairNames.add(ticker.getMarketName()));
        return pairNames;
    }

    private void updateVolumesForMarket(String marketName, List<? extends TickerResponse> tickers) {
        if (marketName.equals(Market.BITTREX_MARKET)) {
            updateVolumes(bittrexVolumes, tickers);
        } else {
            updateVolumes(livecoinVolumes, tickers);
        }
    }

    private void updateVolumes(Map<String, Double> volumeMap, List<? extends TickerResponse> tickers) {
        tickers.forEach(ticker -> volumeMap.put(ticker.getMarketName(), ticker.getVolume()));
    }

    @Override
    public List<String> postMarketProcess(List<String> allMarketNames) {
        return allMarketNames;
    }

    @Override
    public List<String> exceptionProcess(Integer param, String exceptionMessage) {
        return null;
    }

    @Override
    public void doInPostExecute(List<String> allPairNames, ContextHolder holder) {
        PairsHolder pairsHolder = (PairsHolder) holder;
        pairsHolder.setAvailablePairs(excludePairs(allPairNames, holder));
        pairsHolder.setVolumes(bittrexVolumes, livecoinVolumes);
        pairsHolder.removeInvalidPairs();
        pairsHolder.updateAllItems();
    }

    @Override
    public void doInOnCanceled(List<String> result, ContextHolder holder) {
    }

    private List<String> excludePairs(List<String> allPairNames, ContextHolder holder) {
        Context context = holder.getContext();
        List<String> invalidPairs = Arrays.asList(context.getString(R.string.ignored_pairs).split("#"));
        allPairNames.removeAll(invalidPairs);
        List<String> usdPairs = new ArrayList<>();
        allPairNames.forEach(marketName -> {
            if (marketName.startsWith("USD/")) usdPairs.add(marketName);
        });
        allPairNames.removeAll(usdPairs);
        return allPairNames;
    }
}

File: app/src/main/java/com/chess/cryptobot/task/AvailablePairsTask.kt
package com.chess.cryptobot.task

import com.chess.cryptobot.content.ContextHolder
import com.chess.cryptobot.content.pairs.PairsHolder
import com.chess.cryptobot.exceptions.MarketException
import com.chess.cryptobot.market.Market
import com.chess.cryptobot.model.response.TickerResponse

class AvailablePairsTask(pairsHolder: PairsHolder) :
    MarketTask<Int, MutableList<String>?>(pairsHolder) {
    private var availablePairNames: MutableList<String>? = null

    override fun preMarketProcess(param: Int) {}

    @Throws(MarketException::class)
    override fun marketProcess(market: Market, param: Int): MutableList<String>? {
        val tickers = market.getTicker()
        val pairNames = getPairNames(tickers)
        synchronized(this) {
            if (availablePairNames == null) {
                availablePairNames = ArrayList(pairNames)
            } else {
                availablePairNames?.retainAll(pairNames)
            }
        }

        return availablePairNames
    }

    private fun getPairNames(tickers: List<TickerResponse>): MutableList<String> {
        val pairNames: MutableList<String> = ArrayList()
        tickers.forEach { ticker: TickerResponse -> pairNames.add(ticker.tickerName) }
        return pairNames
    }

    override fun postMarketProcess(result: MutableList<String>?): MutableList<String>? {
        return result
    }

    override fun exceptionProcess(param: Int, exceptionMessage: String?): MutableList<String>? {
        return null
    }

    override fun doInPostExecute(result: MutableList<String>?, holder: ContextHolder) {
        val pairsHolder = holder as PairsHolder
        pairsHolder.setAvailablePairs(result)
        pairsHolder.hasAvailablePairs = true
        pairsHolder.removeInvalidPairs()
        pairsHolder.updateAllItems()
    }

    override fun doInOnCanceled(result: MutableList<String>?, holder: ContextHolder?) {}

}